services/replicator/inconn.go

```diff
@@ -75,69 +73,59 @@ func newInConnection(extUUID string, destPath string, stream storeStream.BStoreO
 		destM3Client:         metrics.NewClientWithTags(m3Client, metrics.Replicator, common.GetDestinationTags(destPath, localLogger)),
 		metricsScope:         metricsScope,
 		perDestMetricsScope:  perDestMetricsScope,
-		closeChannel:         make(chan struct{}),
 		creditsCh:            make(chan int32, 5),
 		creditFlowExpiration: time.Now().Add(creditFlowTimeout),
+		shutdownCh:           make(chan struct{}),
 	}
 
 	return conn
 }
 
 func (conn *inConnection) open() {
-	conn.lk.Lock()
-	defer conn.lk.Unlock()
-
-	if !conn.opened {
-		go conn.writeMsgsStream()
-		go conn.readCreditsStream()
-
-		conn.opened = true
-	}
+	conn.wg.Add(2)
+	go conn.writeMsgsStream()
+	go conn.readCreditsStream()
 	conn.logger.Info("in connection opened")
 }
 
-func (conn *inConnection) close() {
-	conn.lk.Lock()
-	defer conn.lk.Unlock()
+func (conn *inConnection) WaitUntilDone() {
+	conn.wg.Wait()
+}
 
-	if !conn.closed {
-		close(conn.closeChannel)
-		conn.closed = true
-	}
-	conn.logger.Info("in connection closed")
+func (conn *inConnection) shutdown() {
+	close(conn.shutdownCh)
 }
 
 func (conn *inConnection) readCreditsStream() {
+	defer close(conn.creditsCh)
+	defer conn.wg.Done()
 	for {
-		select {
-		case <-conn.closeChannel:
+		msg, err := conn.stream.Read()
+		if err != nil {
+			conn.logger.WithField(common.TagErr, err).Info("read credit failed")
 			return
-		default:
-			msg, err := conn.stream.Read()
-			if err != nil {
-				conn.logger.WithField(common.TagErr, err).Info("read credit failed")
-				go conn.close()
-				return
-			}
+		}
 
-			conn.m3Client.AddCounter(conn.metricsScope, metrics.ReplicatorInConnCreditsReceived, int64(msg.GetCredits()))
+		conn.m3Client.AddCounter(conn.metricsScope, metrics.ReplicatorInConnCreditsReceived, int64(msg.GetCredits()))
 
-			// send this to writeMsgsPump which keeps track of the local credits
-			// Make this non-blocking because writeMsgsStream could be closed before this
-			select {
-			case conn.creditsCh <- msg.GetCredits():
-			default:
-				conn.logger.
-					WithField(`channelLen`, len(conn.creditsCh)).
-					WithField(`credits`, msg.GetCredits()).
-					Warn(`Dropped credits because of blocked channel`)
-			}
+		// send this to writeMsgsPump which keeps track of the local credits
+		// Make this non-blocking because writeMsgsStream could be closed before this
+		select {
+		case conn.creditsCh <- msg.GetCredits():
+		case <-conn.shutdownCh:
+			return
+		default:
+			conn.logger.
+				WithField(`channelLen`, len(conn.creditsCh)).
+				WithField(`credits`, msg.GetCredits()).
+				Warn(`Dropped credits because of blocked channel`)
 		}
 	}
 }
 
 func (conn *inConnection) writeMsgsStream() {
 	defer conn.stream.Done()
+	defer conn.wg.Done()
```

Comment:
you should probably have "defer conn.wg.Done()" _before_ "conn.stream.Done()" ..